subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskOutputCachingStateExecuter.java

```diff
@@ -16,36 +16,168 @@
+
+    private final boolean buildCacheEnabled;
     private final TaskExecuter delegate;
 
-    public ResolveTaskOutputCachingStateExecuter(boolean taskOutputCacheEnabled, TaskExecuter delegate) {
-        this.taskOutputCacheEnabled = taskOutputCacheEnabled;
+    public ResolveTaskOutputCachingStateExecuter(boolean buildCacheEnabled, TaskExecuter delegate) {
+        this.buildCacheEnabled = buildCacheEnabled;
         this.delegate = delegate;
     }
 
     @Override
-    public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
-        if (taskOutputCacheEnabled) {
-            TaskOutputCachingState taskOutputCachingState = task.getOutputs().getCachingState(context.getTaskProperties(), context.getBuildCacheKey());
+    public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
+        if (buildCacheEnabled) {
+            TaskOutputCachingState taskOutputCachingState = resolveCachingState(
+                context.getTaskProperties().hasDeclaredOutputs(),
+                context.getTaskProperties().getOutputFileProperties(),
+                context.getBuildCacheKey(),
+                task,
+                task.getOutputs().getCacheIfSpecs(),
+                task.getOutputs().getDoNotCacheIfSpecs(),
+                context.getTaskArtifactState().getOverlappingOutputs()
+            );
+            context.setTaskCachingEnabled(taskOutputCachingState.isEnabled());
             state.setTaskOutputCaching(taskOutputCachingState);
             if (!taskOutputCachingState.isEnabled()) {
                 LOGGER.info("Caching disabled for {}: {}", task, taskOutputCachingState.getDisabledReason());
             }
         } else {
-            state.setTaskOutputCaching(DefaultTaskOutputs.DISABLED);
+            state.setTaskOutputCaching(DISABLED);
+        }
+        return delegate.execute(task, state, context);
+    }
+
+    @VisibleForTesting
+    static TaskOutputCachingState resolveCachingState(
+        boolean hasDeclaredOutputs,
+        Collection<TaskOutputFilePropertySpec> outputFileProperties,
+        TaskOutputCachingBuildCacheKey buildCacheKey,
+        TaskInternal task,
+        Collection<SelfDescribingSpec<TaskInternal>> cacheIfSpecs,
+        Collection<SelfDescribingSpec<TaskInternal>> doNotCacheIfSpecs,
+        @Nullable OverlappingOutputs overlappingOutputs
+    ) {
+        if (cacheIfSpecs.isEmpty()) {
+            return CACHING_NOT_ENABLED;
+        }
+
+        if (!hasDeclaredOutputs) {
+            return NO_OUTPUTS_DECLARED;
+        }
+
+        if (overlappingOutputs != null) {
+            return DefaultTaskOutputCachingState.disabled(TaskOutputCachingDisabledReasonCategory.OVERLAPPING_OUTPUTS,
+                String.format("Gradle does not know how file '%s' was created (output property '%s'). Task output caching requires exclusive access to output paths to guarantee correctness.",
```

Comment:
It would be great to be able to retain a relative path here. At least in Gradle Enterprise this path should be shown as relative.

Reply:
I looked at the corresponding test and I don't see why a relative path makes things more understandable. The user only has the path of the task and a relative path, so if they want to take a look at the actual file they need to do some digging, An absolute path seems more useful to me.

Reply:
Note that the relative path was relative to the project - so it is not hard to find it. An absolute path is very unwieldy, especially when displayed in build scans. I we use an absolute path here, then the build scan side should take care to convert it into a relative path to the root project, if only for deduplication reasons.

@big-guy suggesting using a relative path quite some time ago: https://code-review.gradle.org/cru/REVIEW-6515#CFR-109810